src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8333,6 +8379,312 @@ private StatementSyntax ParseLocalDeclarationStatement()
+            return _syntaxFactory.VariableDeclaration(type, declarator);
+        }
+
+        /// <summary>
+        /// Check ahead for a deconstruction declaration, which requires deconstruction-variables with at least one good-looking variable, followed by an equals sign.
+        /// Doesn't move the cursor.
+        /// PROTOTYPE(tuples) Can this be done without allocations?
+        /// </summary>
+        private bool IsPossibleDeconstructionDeclaration()
+        {
+            if ((this.CurrentToken.IsVar && this.PeekToken(1).Kind == SyntaxKind.OpenParenToken) ||
+                this.CurrentToken.Kind == SyntaxKind.OpenParenToken)
+            {
+                var resetPoint = this.GetResetPoint();
+                try
+                {
+                    var variables = ParseDeconstructionDeclaration(withEquals: false);
+                    var equalsToken = this.EatToken(SyntaxKind.EqualsToken);
+
+                    // We need the equals token and one other confirmation that this is a deconstruction syntax
+                    return DeconstructionVariableLooksGood(variables) && !equalsToken.IsMissing;
+                }
+                finally
+                {
+                    this.Reset(ref resetPoint);
+                    this.Release(ref resetPoint);
+                }
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if one distinct clue is found that this is intended as deconstruction variables
+        /// </summary>
+        private bool DeconstructionVariableLooksGood(VariableDeclarationSyntax node, bool topLevel = true)
+        {
+            if ((object)node == null)
+            {
+                return false;
+            }
+
+            if (node.Type != null && node.Type.Kind == SyntaxKind.IdentifierName && ((IdentifierNameSyntax)node.Type).Identifier.IsVar && node.Deconstruction != null && !node.Deconstruction.OpenParenToken.IsMissing && !node.Deconstruction.CloseParenToken.IsMissing)
+            {
+                // var (..., ....)
+                return true;
+            }
+
+            if (!topLevel && node.Type != null && !node.Type.IsMissing && node.Variables.Count == 1 && !node.Variables[0].Identifier.IsMissing)
+            {
+                // type id
+                return true;
+            }
+
+            if (node.Type == null && node.Variables.Count == 0 && node.Deconstruction != null)
+            {
+                // (..., ...) where one of the elements looks good
+                int count = node.Deconstruction.Variables.Count;
+                for (int i = 0; i < count; i++)
+                {
+                    if (DeconstructionVariableLooksGood(node.Deconstruction.Variables[i], topLevel: false))
+                    {
+                        return true;
+                    }
+                }
```

Comment:
Ah, thanks. It's surprising there's no enumerator on ```InternalSyntax.SeparatedSyntaxList<T>```. Feel free to leave as is then.